addons/sale/models/sale_order.py

```diff
@@ -675,13 +675,12 @@ def _compute_access_url(self):
     @api.constrains('company_id', 'order_line')
     def _check_order_line_company_id(self):
         for order in self:
-            product_company = order.order_line.product_id.company_id
-            companies = product_company and product_company._accessible_branches()
-            if companies and order.company_id not in companies:
-                bad_products = order.order_line.product_id.filtered(lambda p: p.company_id and p.company_id != order.company_id)
+            invalid_companies = order.order_line.product_id.company_id.filtered(lambda c: order.company_id not in c._accessible_branches())
+            if invalid_companies:
+                bad_products = order.order_line.product_id.filtered(lambda p: p.company_id and p.company_id in invalid_companies)
```

Comment:
```suggestion
            invalid_companies = order.order_line.product_id.company_id.filtered(
                lambda c: order.company_id not in c._accessible_branches()
            )
            if invalid_companies:
                bad_products = order.order_line.product_id.filtered(
                    lambda p: p.company_id and p.company_id in invalid_companies
                )
```
Please avoid going over 100 characters on a line.